### PR TITLE
Disease Stages But Epic

### PR DESCRIPTION
**Pull request by EmoGarbage404**
## About the PR 
Stages are now defined as a dictionary of thresholds corresponding to numerically ordered stages starting at 0. Effects and cures can be applied to a stage through the `Stages` datafield, which is simply a list of integer stages.
**Screenshots**

**Changelog**

:cl:
- add: Diseases can now have different stages.

### CODE REVIEW

File: Content.Server/Disease/DiseaseSystem.cs [hunk 1]
```diff
@@ -105,17 +106,29 @@ public override void Update(float frameTime)
...
+                    float lastTreshold = 0;
+                    foreach (var threshold in disease.Stages)
+                    {
+                        if (disease.TotalAccumulator >= threshold.Value &&
+                            threshold.Value > lastTreshold)
+                        {
+                            lastTreshold = threshold.Value;
+                            stage = threshold.Key;
+                        }
+                    }
```
Comment (metalgearsloth) [inline Content.Server/Disease/DiseaseSystem.cs:119]:
```suggestion
                    float lastThreshold = 0;
                    foreach (var threshold in disease.Stages)
                    {
                        if (disease.TotalAccumulator >= threshold.Value &&
                            threshold.Value > lastTreshold)
                        {
                            lastThreshold = threshold.Value;
                            stage = threshold.Key;
                        }
                    }
```

Also should the thresholds just be an array?

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

File: Content.Server/Disease/DiseaseSystem.cs [hunk 2]
```diff
@@ -105,17 +106,29 @@ public override void Update(float frameTime)
...
+                    for (var j = 0; j < disease.Stages.Count; j++)
+                    {
+                        if (disease.TotalAccumulator >= disease.Stages[j] &&
+                            disease.Stages[j] > lastTreshold)
+                        {
+                            lastTreshold = disease.Stages[j];
+                            stage = j;
+                        }
+                    }
+
```
Comment (metalgearsloth) [inline Content.Server/Disease/DiseaseSystem.cs:120]: Still has tresholds

File: Content.Shared/Disease/DiseasePrototype.cs [hunk 3]
```diff
@@ -37,6 +37,19 @@ public sealed class DiseasePrototype : IPrototype, IInheritingPrototype
...
+        /// the total amount of time a disease has been present.
+        /// </summary>
+        public float TotalAccumulator = 0f;
+        /// <summary>
+        /// Stores all the separate stages of the disease plus the time
+        /// thresholds for their activation
+        /// int: the disease stage (0 for baseline, 1, 2, etc.)
+        /// float: the time it takes for the stage to begin.
+        /// </summary>
+        [DataField("Stages", serverOnly: true)]
```
Comment (metalgearsloth) [inline Content.Shared/Disease/DiseasePrototype.cs:50]:
```suggestion
        [DataField("stages", serverOnly: true)]
```

File: Content.Shared/Disease/DiseaseCure.cs [hunk 4]
```diff
@@ -11,7 +11,12 @@ public abstract class DiseaseCure
         /// and false otherwise
         /// </summary>
         public abstract bool Cure(DiseaseEffectArgs args);
-
+        /// <summary>
+        /// What stages the cure applies to.
+        /// probably should be all, but go wild
+        /// </summary>
+        [DataField("stages")]
+        public readonly List<int> Stages = new() { 0 };
```
Comment (metalgearsloth) [inline Content.Shared/Disease/DiseaseCure.cs:19]: Any reason it can't be an array?

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
This needs some sort of validation that the yml is correct.

**Comment by EmoGarbage404** [comment]
I really wasn't sure what to do for testing, but everything else is addressed.

**Review by metalgearsloth — APPROVED** [review]

**Pull request merged**